### ISSUE TITLE
DUP bit set on QoS=0 messages

### ISSUE THREAD

**Issue by yogo1212** [emqtt/emqttd#1319]
#### Environment
- EMQ:
```
# cat releases/start_erl.data
8.3 2.2
```

#### Description
`The DUP flag MUST be set to 0 for all QoS 0 messages` [1].
This causes at least a ruby mqtt gem to throw an error [2].

[1] http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718038
[2] ``/usr/local/rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/mqtt-0.5.0/lib/mqtt/packet.rb:390:in `validate_flags': Invalid packet: DUP cannot be set for QoS 0 (MQTT::ProtocolException)``

**Comment by grutabow** [comment]
@yogo1212   Would you please capture a tcpdump(wireshark) log for the entire procedure? It would help a lot. Thanks!

**Comment by yogo1212** [comment]
@grutabow I will try to set the server up on my computer.
meanwhile: https://github.com/emqtt/emqttd/pull/1323

**Comment by turtleDeng** [comment]
@yogo1212 
qos = 0 message is not redeliver, please provide emqttd log.

**Comment by yogo1212** [comment]
that i can do easily:
```
Exec: /var/emqttd/erts-8.3/bin/erlexec -noshell -noinput +
Root: /var/emqttd
starting emqttd on node 'emqttd@127.0.0.1'
emqttd ctl is starting...[ok]
emqttd hook is starting...[ok]
emqttd router is starting...[ok]
emqttd pubsub is starting...[ok]
emqttd stats is starting...[ok]
emqttd metrics is starting...[ok]
emqttd pooler is starting...[ok]
emqttd trace is starting...[ok]
emqttd client manager is starting...[ok]
emqttd session manager is starting...[ok]
emqttd session supervisor is starting...[ok]
emqttd wsclient supervisor is starting...[ok]
emqttd broker is starting...[ok]
emqttd alarm is starting...[ok]
emqttd mod supervisor is starting...[ok]
emqttd bridge supervisor is starting...[ok]
emqttd access control is starting...[ok]
emqttd system monitor is starting...[ok]
Load emq_mod_presence module successfully.
Load emq_mod_subscription module successfully.
dashboard:http listen on 0.0.0.0:18083 with 2 acceptors.
mqtt:tcp listen on 127.0.0.1:11883 with 16 acceptors.
mqtt:tcp listen on 0.0.0.0:1883 with 16 acceptors.
mqtt:ssl listen on 0.0.0.0:443 with 16 acceptors.
emqttd 2.2 is running now
```
i, too, don't understand why there would be retransmissions for QoS=0.
but i can reproduce it quite reliably.
one way to do it is to start the server, attach a subscriber, detach the subscriber, send a hail of qos=0 messages, and attach same subscriber again.
setting the clean_session flag helps but after a few more qos=0 messages, it happens again.

my attempts to build and run emqtt on my personal computer are failing :-|
and i can't pullt the dump from our prod server for legal reasons.

**Comment by yogo1212** [comment]
no errors or crashes on the server side

**Comment by HJianBo** [comment]
@yogo1212  The problem is caused by the publisher sending a bad message: `Qos = 0` `DUP = 1`

Maybe your question is the same cause?

**Comment by grutabow** [comment]
@yogo1212   You can capture tcpdump(or wireshark) log from subscriber PC, other than server.

**Comment by yogo1212** [comment]
@HJianBo i will check some source packets as well, then.
@grutabow i understand. still, i need an emq server to connect to. right now, i don't have the time set it up.

**Comment by HJianBo** [comment]
@yogo1212 Has fixed and released at v2.3.0
